### PR TITLE
Fix test failure on JDK 15+

### PR DESCRIPTION
**Pull request by DavideD**
Built on top of https://github.com/hibernate/hibernate-reactive/pull/429

### CODE REVIEW

**Comment by DavideD** [comment]
OK,  adding the JDK 8 is not as straightforward as I thought.
I will have a better look tomorrow

**Comment by yrodiere** [comment]
@DavideD @Sanne The default build already uses JDK8. I don't see the point of adding it again in the JDK matrix?

**Comment by yrodiere** [comment]
@DavideD Also, it looks like you're having trouble to add JDK8 because... for some reason Gradle is producing Java 11 bytecode. I'm investigating.

**Comment by yrodiere** [comment]
@DavideD Got it. I fixed the problem and added a build that tests with JDK8 to the JDK matrix. It's probably a waste of resources, but I suppose we've got plenty at the moment :)

**Comment by yrodiere** [comment]
(@DavideD Beware: I force-pushed to your branch)

**Comment by DavideD** [comment]
@yrodiere That won't work

**Comment by DavideD** [comment]
OK, forget what I said. 

My comment refers to the fact that I'm sure yesterday I've sent the same change and the build failed.
But it was late and I suspect my commit wasn't exactly the same.

Thanks @yrodiere

**Comment by yrodiere** [comment]
@DavideD As discussed, I pushed a few more commits:

1. When testing against JDK15 for example, we don't just run tests with JDK15: we also compile tests with JDK15 and use `-release 15`. That way, we really simulate an application using JDK15 for everything, while relying on our own binaries built for JDK8.
2. When testing against JDK 11 / 15 / 16, we now build the main code with JDK 8, not 11. That way, the produced binaries are identical to what we will actually release.
3. Simplify the Gradle code. Not strictly necessary, but I think it's clearer that way, and more importantly it will be easier to adapt in the future. That's important because I intend to port this code to ORM (next week).

I think @Sanne will be pleased with item 2 in particular.

File: build.gradle [hunk 1]
```diff
@@ -103,9 +113,29 @@ subprojects {
     version = projectVersion
 
     if ( project.javaToolchainEnabled ) {
-        compileJava.options.release = project.baselineJavaVersion.asInt()
-        compileTestJava.options.release = project.baselineJavaVersion.asInt()
-        compileJava.options.encoding = 'UTF-8'
+        // "sourceCompatibility" is not supported with toolchains. We have to work around that limitation.
+        compileJava.configure {
+            if (project.compileJdkVersion.asInt() < 11) {
```
Comment (DavideD) [inline build.gradle]: Wasn't `release` introduced in 9?

File: build.gradle [hunk 2]
```diff
@@ -59,20 +59,103 @@ ext {
...
+            } else {
+                options.release = project.javaVersions.test.release.asInt()
+            }
+            options.encoding = 'UTF-8'
+        }
+    }
+    else {
+        sourceCompatibility = JavaVersion.toVersion( project.baselineJavaVersion )
+        targetCompatibility = JavaVersion.toVersion( project.baselineJavaVersion )
+        compileJava.options.encoding = 'UTF-8'
```
Comment (DavideD) [inline build.gradle:157]: Can't we just set this option once before the initial if?

**Comment by DavideD** [comment]
LGTM

**Comment by DavideD** [comment]
Really nice

File: build.gradle (hunk 1 above)
Comment (yrodiere) [reply build.gradle]: It was. I added a commit. We don't test Java 9 / 10 though: they are not supported anywhere AFAIK.

File: build.gradle (hunk 2 above)
Comment (yrodiere) [reply build.gradle]: Right. I amended my last commit and force-pushed.

File: build.gradle (hunk 1 above)
Comment (DavideD) [reply build.gradle]: :)
Thanks, I was just checking, I didn't mean that you should fix it.

**Comment by DavideD** [comment]
Merged https://github.com/hibernate/hibernate-reactive/pull/430

**Comment by Sanne** [comment]
this is great! thanks again

**Pull request merged**